### ISSUE TITLE
[0.54] [iOS] Wrong 'Backspace' event fired onKeyPress after clearing TextInput

### ISSUE THREAD

**Issue by michalpetrov** [facebook/react-native#18374]
In IOS, after clearing TextInput value (setting it to empty string) and then pressing any key (e.g. 'a') two onKeyPress events are fired: A, Backspace. First is for the pressed key and second is a Backspace.

### Environment

Environment:
  OS: macOS Sierra 10.12.6
  Node: 8.9.1
  Yarn: 1.3.2
  npm: 5.5.1
  Watchman: 4.9.0
  Xcode: Xcode 9.2 Build version 9C40b
  Android Studio: 3.0 AI-171.4443003

Packages: (wanted => installed)
  react: ^16.3.0-alpha.1 => 16.3.0-alpha.1
  react-native: 0.54.2 => 0.54.2

### Expected Behavior

Only one event for the pressed key should be fired (e. g. 'A').

### Actual Behavior

An additional 'Backspace' event is fired.

### Steps to Reproduce

1. write anythig in the textfield
2. press clear
3. press 'a' >>> events 'A' and 'Backspace' will be fired

```javascript
type Props = {};
type State = {
  text: string,
  keys: string
};
export default class App extends Component<Props, State> {
  state = {text: '', keys: ''}
  render() {
    return (
      <View style={styles.container}>
        <TextInput style={styles.textInput} value={this.state.text} onChangeText={this.onChangeText} onKeyPress={this.onKeyPress}/>
        <Button title="Clear" onPress={this.onClear}/>
        <Text>Text: {this.state.text}</Text>
        <Text>Keys: {this.state.keys}</Text>
      </View>
    );
  }

  onChangeText = (text: string) => {
    this.setState({text})
  }

  onKeyPress = ({ nativeEvent }: Object) => {
    this.setState({keys: this.state.keys + nativeEvent.key + ', '})
  }

  onClear = () => {
    this.setState({text: '', keys: ''})
  }
}
```

**Comment by johnckendall** [comment]
+1, validated this is happening

**Comment by edo1493** [comment]
+1

**Comment by choiRyan** [comment]
I also encountered this. It causes react-native-tag-input@0.0.18 to not work properly. (Used the example code in the repo for replication) https://github.com/jwohlfert23/react-native-tag-input

**Comment by hamaron** [comment]
+1; I've got the same issue on 0.54.

**Comment by barsumek** [comment]
+1, encountered when using https://github.com/jwohlfert23/react-native-tag-input

**Comment by hamaron** [comment]
I opened a pull request to fix this bug: #18627

**Comment by kpatt1011** [comment]
+1 to this happening in our app where we are validating a Confirmation Code.

When can we get #18627 merged in?

**Comment by kayzenkayzen** [comment]
+1

**Comment by Chloro** [comment]
+1

**Comment by sgehly** [comment]
+1

**Comment by sgehly** [comment]
In the meantime, I fixed this in my confirmation code page by running .clear() on the input before setting the state to "".

```
this.refs["input"].clear()
this.refs["input"].focus()
this.setState({value:""})
```

I don't know how it worked, but it did.

**Comment by kayzenkayzen** [comment]
Not worked for me!

**Comment by Arthur-ZB** [comment]
+1

**Comment by alxyuu** [comment]
+1, experience this on 0.55.2 as well

**Comment by amarw** [comment]
I was able to workaround this bug by adding a check in the `onKeyPress` callback function, 

```
 onKeyPress(e) {
    if (e.nativeEvent.key === 'Backspace') {
      // Return if duration between previous key press and backspace is less than 20ms
      if (Math.abs(this.lastKeyEventTimestamp - e.timeStamp) < 20) return;

      // Your code here
 
    } else {
      // Record non-backspace key event time stamp
      this.lastKeyEventTimestamp = e.timeStamp;
    }
  }
```

I found while debugging that duration between receiving backspace event after key press is less than ~10ms. So I record the non-backspace key event timeStamp and use it to compare with backspace event timeStamp. If the difference between these two is less than 20ms (arbitrary value > 10ms) then I return from the callback. 
Hope this helps!

**Comment by hramos** [comment]
Fixed.

**Comment by Ashoat** [comment]
Fix should be in `react-native@0.56.0-rc.2`!

**Comment by michaelkowalski** [comment]
@amarw it's event better to check nativeEvent.eventCount (at least in 0.55.2), they are the same in the two events :)
